### PR TITLE
xmlchange: Give user guidance on what to do next

### LINKED ISSUE #815: Feature request: xmlchange tells you what steps need to be rerun for the change to take effect
This is a feature request that popped to mind, which is for the sake of relatively inexperienced users: It would be nice if xmlchange noticed when you change a setting that requires you to re-do some step, and gave you the commands you need to run for the change to take effect.

For example, if you change a setting in env_build.xml after you have already done the build, then it will tell you something like:

```
For this change to take effect, run:
./case.build --clean-all
./case.build
```

(I'm actually not sure if those are the commands you need to run when changing env_build.xml... maybe this feature request is as much for me as for "inexperienced users", or maybe I myself fall into the "inexperienced user" category in this respect....)

I'm imagining some pretty simple rules like: If a setting in env_build is changed after the build is complete, then you need to redo the build; if a setting in env_mach_pes is changed after case.setup is complete, then you need to redo case.setup. That may be all that's really needed.

What made me think of this was thinking of how things have changed since CESM1: In CESM1, you needed to explicitly specify the file to change via xmlchange. Now that is determined automatically. That automatic determination is very nice, but the downside is that it lets you change files without realizing that (for example) you're changing a build-time setting that requires a rebuild.

### PR DESCRIPTION
**Pull request by jgfouca**
Test suite: by-hand
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #815 

User interface changes?: N

Update gh-pages html (Y/N)?:

Code review: @jedwards4b @billsacks

### CODE REVIEW

**Comment by jgfouca** [comment]
@billsacks , approve?

File: scripts/lib/CIME/case/case.py [hunk 1]
```diff
@@ -368,11 +368,11 @@ def get_resolved_value(self, item, recurse=0, allow_unresolved_envvars=False):
 
         return item
 
-    def set_value(self, item, value, subgroup=None, ignore_type=False, allow_undefined=False):
+    def set_value(self, item, value, subgroup=None, ignore_type=False, allow_undefined=False, return_file=False):
         """
         If a file has been defined, and the variable is in the file,
-        then that value will be set in the file object and the file
-        name is returned
+        then that value will be set in the file object and the resovled value
```
Comment (billsacks) [inline scripts/lib/CIME/case/case.py]: Please document the return value(s) - i.e., if return_file is False, then just the resolved value is returned; if True, then a tuple (value, filename) is returned. If the variable isn't found, then returns None.

File: scripts/Tools/xmlchange [hunk 2]
```diff
@@ -143,8 +143,19 @@ def xmlchange_single_value(case, xmlid, xmlval, subgroup, append, force, dryrun)
...
+        expect(result is not None,"No variable \"%s\" found"%xmlid)
+        filename = result[1]
+
+        if filename.endswith("env_batch.xml"):
+            logger.info("case.submit will auto-regen your batch files, no further user action is needed")
+        elif filename.endswith("env_build.xml"):
+            logger.info("You should consider cleaning and rebuilding your case")
+        elif filename.endswith("env_mach_pes.xml"):
+            logger.info("Changing mach_pes can have a big impact on your case. Please re-run case.setup, and clean/rebuild your case")
+        elif filename.endswith("env_case.xml"):
```
Comment (billsacks) [inline scripts/Tools/xmlchange]: I thought you're not allowed to change something in env_case, ever. Is it really safe to change these values if you rerun case.setup and rebuild?

Comment (jgfouca) [reply scripts/Tools/xmlchange]: I was wondering that myself. @jedwards4b ?

**Comment by jgfouca** [comment]
@billsacks , your "out of order" is a great idea. Let me work in this PR a bit more.

File: scripts/Tools/xmlchange (hunk 2 above)
Comment (jedwards4b) [reply scripts/Tools/xmlchange]: Right - you should never change anything in env_case.xml -  I think there is a way to mark that file as read only.    There are a few special circumstances where you may want to change something in this directory but I'm not sure it'll work - say for example you need to move a case directory.

**Comment by jgfouca** [comment]
@billsacks , see latest changes.

**Comment by rljacob** [comment]
I like the change to only print something if its out of order.

**Comment by jedwards4b** [comment]
```
./xmlchange ROOTPE_CPL=0
For your changes to take effect, run:
./case.setup --reset
./case.build --clean-all
./case.build
```
Is overkill - the --clean-all is not needed and makes the rebuild quite slow

**Comment by billsacks** [comment]
@jedwards4b so what do you suggest instead?

**Comment by jedwards4b** [comment]
```
For your changes to take effect, run:
./case.setup --reset
./case.build
```

**Comment by jedwards4b** [comment]
The exception is if there is a change from no threading to threading

**Pull request merged**